cpp/src/Slice/Parser.h

```diff
@@ -158,14 +162,61 @@ namespace Slice
         virtual bool shouldVisitIncludedDefinitions() const { return false; }
     };
 
+    // ----------------------------------------------------------------------
+    // GrammarBase
+    // ----------------------------------------------------------------------
+
+    class GrammarBase
+    {
+    public:
+        virtual ~GrammarBase() = default;
+    };
+
+    // ----------------------------------------------------------------------
+    // Metadata
+    // ----------------------------------------------------------------------
+
+    class Metadata final : public virtual GrammarBase
+    {
+    public:
+        Metadata(const std::string& rawMetadata, const std::string& file, int line);
+        std::string_view directive() const;
+        std::string_view arguments() const;
+
+        std::string file() const;
+        int line() const;
+
+        friend std::ostream& operator<<(std::ostream& out, const Metadata& metadata);
+
+    private:
+        /// Parses a metadata string into a pair of (directive, arguments) strings.
+        static std::pair<std::string, std::string> parseRawMetadata(const std::string& rawMetadata);
```

Comment:
string_view?

Reply:
string.

Now the caller just forwards a string by moving it.
This lets us avoid a copy in the case where we don't want to split up the metadata at all,
and we can just return the string as-is, instead of needing to construct a new one from the `string_view`.

Reply:
So the parameter should be `std::string`, not ` `const std::string&`? That's fine too.

Reply:
I went with a 3rd compromise option to save time debating here: I inlined the function.

I agree with you, but it didn't seem like we'd be able to agree on the intent of the parameter here.
I thought that by-value `string` (that it's free to modify however) was correct, but it's clear no one else did.

And when I switched it to `string_view`, it just hurt me too much to look at.
It was a private static helper function, called from only one place, that will be passed a `string` 100% of the time.
If switched to `string_view`, the first line copied it into a string anyways `string x = string{my_string_view}`.